README

Seek to the section ``INSTALLATION'' if you want the quickie.


PROJECT OVERVIEW

	I started ident2 awhile ago for several reasons.
		A: I needed something to do.
		B: I wanted to make my reply something other than my
			generic UNIX username, for IRC chaos, I guess.
		C: I needed the practice.
		D: I wanted to contribute to the community!

	I've since grown weary of meddling with this server.  After the
	second rewrite between a feature incorporation blitz and a
	minimalization streak, I've decided to move on.  I'll still add
	bug fixes and address any security concerns, however.

TERMS OF USE

	Ident2 is provided under the GNU General Public License.
	This license can be viewed in the included 'COPYING' file or
	found at http://www.gnu.org/


PLANNED PORTS

	I am very much in favor of porting this server. I have hopefully
	complied with POSIX well enough to make it reasonably simple.
	Writing ports should be a no brainer as far as writing actual code
	goes. The only parts that need implementing are the three m_*
	functions listed in the MACHINE section in ident2.h

	Ident2 was developed under Linux and has been ported to FreeBSD.


PORT NOTES

	The FreeBSD port will probably only compile under FreeBSD 3.x and up
        The Linux version was tested only on 2.2.x


GETTING

	https://github.com/mbacarella/ident2

INSTALLATION

	Check define.h first for things you might be interested in toggling.

	./configure
	make
	make install

	If the script fails, your system is probably not supported or
	you have a buggered install. I could've also screwed up.
	If you think it's MY fault, e-mail me.
		(seek to the section CONTACT, below)

	The server should install to /usr/local/sbin/ident2

	Start the server from the command line (as root), or
	go right ahead and add it to /etc/inetd.conf if you're too
	cool for your own good. If you have no idea how to do this,
	please read the man page. ``man inetd.conf'' works nicely.
	Here is what you might use:

auth    stream  tcp     nowait  root    /usr/local/sbin/ident2  ident2

	Yes, these fields are deliciously tab seperated.
	If inetd.conf has no idea what 'auth' is, make sure
	it appears in /etc/services. If not, well, you're going
	to learn how to do two new things today.

	Feel free to also spawn ident2 with tcp wrappers. It shouldn't mind.

	If you're sick of the damned thing, you can uninstall it by
	doing:
		make uninstall

	Be sure to remove it from inetd.conf if necessary.

IP MASQUERADING NOTES

        IP Masquerading support was removed. I do not feel that I can
	implement it without disgustifying my code, but furthermore, I
	don't have the time/patience to make it work.

	You can essentially achieve the same results by telling ident2
	to simply send random replies to all requests.

CONFIGURING REPLIES

	By default, without any command line options,
	ident2 should be fully usable in production environments
	that require user accountability. Divergence from the RFC
	is only achieved by command line parameters that tell it
	to do so.

	In short, if you're looking for a drop in replacement,
	just start ident2 from the command line (as root) with
	nothing else.

	If you're wondering how to set your own ident replies,
	merely create a file .ident in your home directory and
	do the following:
		echo 'ident spanky' > ~/.ident

	Start ident2 with the '-i' option. You may of course, replace
	'spanky' with whatever your preffered ident reply might be.

	Also note. If a user has the file ``.noident'' in their directory,
	replies will be disabled. You must start the server with '-n' if
	you want this feature.

MODES OF OPERATION

	Ident2 has two modes of operation. inetd child, or
	standalone daemon. The inetd child itself, never forks.
	The daemon fork()s on every accepted connection.

	Whatever the mode, execution must trickle down to the nexus
	if anything useful is supposed to happen. The function nexus()
	is basically where the logic occurs and the reply is decided on
	and dispatched.

CONTACT

	Michael Bacarella <mbac@panix.com>
	http://panix.com/~mbac/


CHANGES
--------- [07/22/2005
+ incorporated lsof patch which adds support for
  SunOS, NetBSD, OpenBSD, OSF/1, HP-UX
  (thanks inittab@netizenweb.com (Greg Schenzel))

--------- [05/18/2004
+ buffer overflow in _getl fixed :(

--------- [11/02/2003
+ fixed using data after it has been freed in sys/m_fbsd.c:m_get_uid()
  thanks to Ilya Novoselov <nullguid@t72.ru>

--------- [08/18/2000
+ forgot to specifiy option 'n' in getopt_long
  thanks to Johan Ekenberg <johan@ekenberg.se> for pointing this out

--------- [04/11/2000
+ correctly handles errno == EINTR when accept() returns -1
+ date formats in CHANGES file now Y2K compliant!

--------- [11/2/99
+ port to FreeBSD
+ PID registration, under FreeBSD, is also /var/run

--------- [9/18/99
+ changed all sprintf's to snprintf's
+ removed all potential security hazards (strcpy, memcpy, sprint, etc)
+ moved m_register_pid AFTER the initial fork()-to-become-daemon,
  otherwise wrong pid is registered.
+ d'oh. move m_reduce_rights AFTER m_register_pid, otherwise can't
  write pid to /var/run
- NOTE, snprintf is not portable. must be removed in future.

--------- [9/3/99]
+ performing security sweep. reevaluating choices
+ merged getopt_long support
+ added getopt_long detection to ./configure
+ wrote two help outputs (short and short+long versions)
+ implemented the --license parameter
+ implemented the --usage paramater

--------- [8/30/99]
+ moved srand(time(NULL)) out of INIT_GLOBALS and into the random reply generator
  function. rand is seeded every time, which is flawed, but still harmless.
  this was moved because the parent never actually calls rand(), which sucks
  because the children will keep generating the same reply.
+ implemented new machine call, m_register_pid(), for use in daemon mode,
  on systems that wish to keep track of running system services.
+ PID registration, under Linux, in /var/run
+ switched to a ./configure based setup

--------- [8/25/99]
+ changed m_linux.c to only setuid(). Under Linux, setuid() sets the effective
  userid as well as superuser. If this would be the case, seteuid() should fail.

--------- [8/9/99]
+ added random replies
+ fixed standard unspecial replies (INIT_GLOBALS wasn't initting!)
- ripped out config file parsers
- ripped out all semblances of IP masquerading
+ rather than trying to drop userid to nobody and failing if it
  doesn't work, it only tries to change uid if uid is 0, elsewise,
  keep current uid.
+ reworded README
  
--------- [6/11/99]
- removed throttled daemon. I felt this included some very unnatural
  code and it was a pain to debug. it was a research project at most.
+ rewrote config file parser
+ rewrote command-line processor
+ included GNU information on top of all source files
+ code cleaning spree. Optimized, minimalized, etc.
+ fixed buffer handling in some areas which may have been dangerous
+ reorganized some general code into common.c
+ stablized threaded daemon
+ rewrote README
+ reorganized throttled daemon
+ moved accept() out into the child process so we don't have
  to do any nasty stuff like return file descriptors to the
  parent through exit().
+ adopted capitalize global variables convention
+ decided to start DATING these changes

---------- 0.999c
+ no idea why I thought the user of 'nobody' couldn't carry on identity
  functions. The dropping to nobody under Linux has been reenacted.
+ fixed a diverge from RFC1413. Thanks to Piotr Kucharski <chopin@sgh.waw.pl>
  for pointing out that END OF LINE is clearly defined by the RFC as CRLF
  and not LF. Apologies to anyone affected by this.
+ fixed an 'off by one' error in the throttled implementation causing a
  garbage char to appear at the end of an ident query.
+ command-line option to display compile-time options.
+ shamelessly copied a pidentd convention. If you have the file .noident
  in your home directory, it will not send a response to anyone requesting
  an ident that resolves to your username.
+ dropping userid to 'nobody' has been changed to instead drop the access
  level to a user defined variable (which SHOULD be a username).
+ moved quite a few variables out into seperate #defines. 
+ dropping userid to 'nobody' as a security measure is back, again.
+ IP masquerading works! Much love to everyone who helped the
  debugging/fixing process. Especially Hans.
  -Actually, it only sort of works. Linux 2.0 and 2.1 have
  differing levels of success. Kernel 2.2 is supposedly going to
  settle on a format once and for all.
+ rand() wasn't being seeded. 
+ for some reason I wasn't making use of #define USER_REPLIES.
+ dropping the userid to 'nobody' makes servicing ident requests
  impossible. I'm dumb. The function 'go_nobody' and all references
  to it were removed.
+ major MAJOR rewrite. This introduced some more natural (ie, not kludgy)
  code as well as more bugs! It should also be easier to extend and
  expand. Hopefully noone will get hurt this time. :)



common.c
/*
 * Ident-2 - an Identity server for UNIX
 * Copyright (C) 1998-2001 Michael Bacarella
 * Copyright (C) 2003 Netgraft Corporation
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 * Please view the file README for program information.
 *
 *	Common function definitions.
 *	Used without regulation by other modules
 */

#include "ident2.h"

void *
xmalloc (size_t nb)
{
	void *p = malloc (nb);
	
	if (p == NULL) {
		fprintf (stderr, "ident2: terminating on memory allocation error\n");
		syslog (LOG_ERR, "error: memory allocation error");
		exit (1);
	}
	return p;
}

	/*
	 *	a (skewed) fgets() that works on file descriptors
	 *	the '\r' charecter is ignored
	 *	returns the number of bytes written into the given
	 *	 buffer, including the terminating NUL
	 */
static int
_getl (int d, char *begin, u_short len)
{
	char *p, *end;

	end = &begin[len-1]; /* leave room for terminating NUL */
	for (p = begin; p < end; ++p) {
		if (read (d, p, 1) != 1)
			break;
		if (*p == '\n')
			break;
		if (*p == '\r')
			p--;	/* ignore \r */
	}
	*p++ = 0;
	return p-begin;
}

	/*
	 *	this function is called directly as an inetd
	 *	handler as well as a fork()ed into by the daemon
	 *	service of the server..
	 */
void 
child_service (int in, int out)
{
	char buf[ID_BUF_SIZE];

	signal (SIGALRM, exit);
	alarm (Client_Timeout);

	if (m_reduce_rights () == -1)
		syslog (LOG_ERR, "error: cannot reduce self's rights "
			"[m_reduce_rights()]");
	else
		while (_getl (in, buf, ID_BUF_SIZE) != 0)
			nexus (out, buf);

	close (in);
	if (out != in)
		close (out);
}


typedef struct cl_t {
	int			sd;
	uid_t			uid;
	unsigned short		lp, rp;
	struct sockaddr_in	laddr, raddr;
	struct passwd		*pw;
} cl_t;

	/*
	 *      send an ident formatted reply to client
	 */
static void
_clreply (cl_t *cl, char *rslt, char *info)
{
	char buf[ID_BUF_SIZE+1];

	buf[ID_BUF_SIZE] = 0;
	
	snprintf (buf, ID_BUF_SIZE-1, "%d , %d : %s : %s\r\n",
		cl->lp, cl->rp, rslt, info);
	write (cl->sd, buf, strlen (buf));
}


static int
_check_noident (cl_t *cl)
{
	char p[ID_BUF_SIZE+1];
	
	p[ID_BUF_SIZE] = 0;
	
	snprintf (p, ID_BUF_SIZE-1, "%s/%s", cl->pw->pw_dir, NOIDENT_FILE);
	
	if (access (p, F_OK) == 0) {
		_clreply (cl, "ERROR", "HIDDEN-USER");
		syslog (LOG_INFO, "no reply to %s (query %d, %d) due to %s",
				inet_ntoa (cl->raddr.sin_addr),
				cl->lp, cl->rp, p);
		return 0;
 	}
	return -1;
}

static int
_check_user_ident (cl_t *cl)
{
	FILE *fp;
	int retval = -1;
	char idfile[ID_BUF_SIZE+1];
	
	idfile[ID_BUF_SIZE] = 0;
	snprintf (idfile, ID_BUF_SIZE-1, "%s/%s",
		cl->pw->pw_dir, User_Ident_File);
	
	if (access (idfile, R_OK) == 0) {
		if ((fp = fopen (idfile, "r")) != NULL) {
			char *p, buf[4096];
			
			memset (buf, 0, 4096);	
			fgets (buf, 4095, fp);
			strtok (buf, "\r\n");
			
			if ((p = strstr (buf, "ident "))) {
				char rply[ID_BUF_SIZE+1];
				rply[ID_BUF_SIZE] = 0;
				
				snprintf (rply, ID_BUF_SIZE-1, "UNIX : %s", p+6);
				_clreply (cl, "USERID", rply);
				
				syslog (LOG_INFO, "sent reply `%s' to query %s "
					"(%d, %d), uid = %d", p,
					inet_ntoa (cl->raddr.sin_addr),
					cl->lp, cl->rp, cl->uid);
				retval = 0;
			}
			fclose (fp);		
		}
	}
	return retval;
}	

	/*
	 * handy for users who use an ip masqueraded setup that just
	 * want to friggin get on IRC and don't care what their ident
	 * is. Or if you want to annoy your IRC admins, because they
	 * totally deserve it.
	 */
static void
_send_random_reply (cl_t *cl)
{
	char randstr[RAND_STRING_LENGTH+1];
	char buf[ID_BUF_SIZE+1];
	size_t i;

	buf[ID_BUF_SIZE] = 0;
	
	srand (time(NULL));
	for (i = 0; i < RAND_STRING_LENGTH; i++) {
		randstr[i] = rand();
		while (randstr[i] > 'z')
			randstr[i] -= 26;
		while (randstr[i] < 'a')
			randstr[i] += 26;
	}
	randstr[RAND_STRING_LENGTH] = 0;
 	
 	snprintf (buf, ID_BUF_SIZE-1, "UNIX : %s", randstr);

	_clreply (cl, "USERID", buf);
	syslog (LOG_INFO, "sent random reply for query %s (%d, %d)",
		inet_ntoa (cl->raddr.sin_addr), cl->lp, cl->rp);
	
	return;
}
	/*
	 *	determines and collects client information
	 *	and returns it in a neat and conveniant package
	 */	
static cl_t *
_new_cl (int sd, char *line)
{
	int uid, ssiz = sizeof (struct sockaddr);
	cl_t *p, cl;
	char *s;
	
	cl.sd = sd;
	cl.lp = cl.rp = 0;
	
	if (!(s = strchr (line, ','))) {
		_clreply (&cl, "ERROR", "INVALID-PORT");
		return NULL;
	}		
	*s = 0;
	cl.lp = (u_short) atoi (line);
	cl.rp = (u_short) atoi (s+1);

	if (getsockname (sd, (struct sockaddr *)&cl.laddr, &ssiz) == -1) {
		syslog (LOG_WARNING, "warning: getsockname(): %s\n",
			strerror (errno));
                _clreply (&cl, "ERROR", "UNKNOWN-ERROR");
                return NULL;
	}
	if (getpeername (sd, (struct sockaddr *)&cl.raddr, &ssiz) == -1) {
		syslog (LOG_WARNING, "warning: getpeername(): %s\n",
			strerror (errno));
		_clreply (&cl, "ERROR", "UNKNOWN-ERROR");
		return NULL;
        }

	if (Reply_Always_Random == TRUE) {
		_send_random_reply (&cl);
		return NULL;
	}

	uid = m_get_uid (&cl.laddr.sin_addr, cl.lp, &cl.raddr.sin_addr, cl.rp);
	if (uid == -1) {
		syslog (LOG_WARNING, "warning: bad request: %d/%d, from %s",
			cl.lp, cl.rp, inet_ntoa (cl.raddr.sin_addr));
		_clreply (&cl, "ERROR", "NO-USER");
		return NULL;
	}
	cl.uid = uid;
		
	if ((cl.pw = getpwuid (cl.uid)) == NULL) {
		syslog (LOG_ERR, "warning: cannot map %d to a username: %s "
			"requested by query %s (%d, %d)", cl.uid,
				strerror (errno), inet_ntoa (cl.raddr.sin_addr),
				cl.lp, cl.rp);
		_clreply (&cl, "ERROR", "UNKNOWN-ERROR");
		return NULL;
	}
	p = xmalloc (sizeof (cl_t));
	*p = cl;
	
	return p;				
}

	/**
	 **	nexus
	 **
	 **	this is called by all service modes if
	 **	any reply is to be returned to the
	 **	client.
	 **/
void
nexus (int sd, char *line)
{
	cl_t *cl;
	char buf[ID_BUF_SIZE+1];

	buf[ID_BUF_SIZE] = 0;
	
	if ((cl = _new_cl (sd, line)) == NULL)
		return;

		/* if user configs are allowed, use the user's (if they
		 * have one, else just use the global config */
	if (Allow_NOIDENT == TRUE)
		if (!_check_noident (cl)) {
			free (cl);
			return;
		}
		
	if (Use_User_Ident == TRUE)
		if (!_check_user_ident (cl)) {
			free (cl);
			return;
		}
	
	snprintf (buf, ID_BUF_SIZE-1, "UNIX : %s", cl->pw->pw_name);
	
	_clreply (cl, "USERID", buf);
	syslog (LOG_INFO, "sent reply `%s' to query %s "
			"(%d, %d), uid = %d", cl->pw->pw_name,
			inet_ntoa (cl->raddr.sin_addr), cl->lp,
			cl->rp, cl->uid);

	free (cl);
	return;		
}



daemon.c
/*
 * Ident-2 - an Identity server for UNIX
 * Copyright (C) 1998-2001 Michael Bacarella
 * Copyright (C) 2003 Netgraft Corporation
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 * Please view the file README for program information.
 *
 * ---------------------
 *	This file provides the 'daemon' implementation of
 *	ident2.
 *
 *	Daemon waits for connections
 *	forks, handles connections, cleans up,
 *	dies, etc.
 *
 *	Module entry point is ``daemon_service''
 * ---------------------
 */

#include "ident2.h"


static int
_go_daemon (void)
{
	switch (fork()) {
	case 0:
		setsid();
		m_register_pid ();
		m_reduce_rights ();
		return 0;
	case -1:
		syslog (LOG_ERR, "d_core: error fork(): %s\n", strerror(errno));
		return -1;
	default:
		exit (0);
	}
}
	
	/*
	 *	remove dead children from the process vector
	 */
static void
_reap_proc (pid_t *pv, size_t Max_Connections)
{
	pid_t *p;
	int r, s;

	for (p = pv; p < pv+Max_Connections; p++) {
		if (*p == 0)
			continue;

		while ((r = waitpid (*p, &s, WNOHANG)) == -1)
			if (errno != EINTR) {
				syslog (LOG_NOTICE,
					"warning: waitpid() error: %s",
						strerror (errno));
				break;
			}
		if (r == 0)
			continue;
		if (r > 0)
			*p = 0;
	}
}


static void
declient (int s)
{
	struct sockaddr sin;
	int ss = sizeof (sin);
	close (accept (s, (struct sockaddr *)&sin, &ss));
}

                                                 	
static void _sig_ign (int s) { return; }


static int
_accept_connect (int sv, struct sockaddr_in *sin)
{
	size_t sl = sizeof (struct sockaddr_in);
	int cl;

	while ((cl = accept(sv, (struct sockaddr *)sin, &sl)) == -1) {
		if (errno == EINTR)
			continue;
		return -1;
	}
	return cl;
}
	
	/*
	 *	will wait for a connection on SV,
	 *	find a space in the proc table and fork() 
	 *	child for it. the child will accept() it and
	 *	process the waiting client.
	 *	we rely on the delivery of EINTR to interrupt
	 *	select(), which gives us a chance to test for
	 *	deceased children.
	 */
static void
d_core (int sv)
{
	struct sigaction sa;
 	struct sockaddr_in sin;
 	pid_t	*pv;
				/* become daemon */
	if (_go_daemon () == -1) {
		syslog (LOG_ERR, "d_core: error fork(): %s\n", strerror(errno));
		return;
	}
			
	pv = xmalloc (Max_Connections * sizeof (pid_t)); 
	memset (pv, 0, Max_Connections * sizeof(pid_t));
	 
	sa.sa_handler = _sig_ign;
	sigemptyset (&sa.sa_mask);
	sa.sa_flags = 0;
	
	if (sigaction (SIGCHLD, &sa, NULL) == -1) {
		syslog (LOG_ERR, "error: registering SIGCHLD handler: %s",
			strerror (errno));
		return;
	}
	
	while (1) {
		pid_t *p;		
		int cl;
		fd_set rfd;
		
		FD_ZERO(&rfd);
		FD_SET(sv, &rfd);
		
		if (select (sv+1, &rfd, NULL, NULL, NULL) == -1) {
			if (errno != EINTR)
				break;
			_reap_proc (pv, Max_Connections);
			continue;
		}
		if (!FD_ISSET (sv, &rfd))
			continue;

		for (p = pv; p < pv+Max_Connections; p++)
			if (*p == 0)
				break;
			
		if (p == pv+Max_Connections) {
			syslog (LOG_INFO, "refusing %s: too many open "
				"connections", inet_ntoa (sin.sin_addr));
			declient (sv);
		}
	
		if ((cl = _accept_connect (sv, &sin)) == -1) {
			syslog (LOG_NOTICE, "warning: inconsistency error: "
				"select() and accept() disagree: %s",
					strerror (errno));
			continue;
		}	
		
		if ((*p = fork()) == -1) {
			syslog (LOG_ERR, "warning: couldn't fork(): %s",
				strerror (errno));
			*p = 0;
		} else if (*p == 0) {
			child_service (cl, cl);
			close (cl);
			exit (0);
		}
		close (cl); /* close now that it's mapped into the child */
	}
}


void daemon_service (void)
{
	int s;
	struct sockaddr_in sin;

	openlog ("ident2", LOG_PID, LOG_DAEMON);
		
	if ((s = socket (PF_INET, SOCK_STREAM, 0)) == -1) {
		syslog (LOG_ERR, "error: socket(): %s", strerror (errno));
		return;
	}
	
	sin.sin_family = AF_INET;
	sin.sin_port = htons (Ident_Port);
	sin.sin_addr.s_addr = INADDR_ANY;

	if (bind (s, (struct sockaddr *)&sin, sizeof (sin)) == -1) {
		syslog (LOG_ERR, "error: binding to port %d: bind(): %s",
			Ident_Port, strerror (errno));
		fprintf (stderr, "error: binding to port %d: bind(): %s\n\n",
			Ident_Port, strerror (errno));
		return;
	}
	
	if (listen (s, 15) == -1) {
		syslog (LOG_ERR, "error: listening to port %d: listen(): %s",
				Ident_Port, strerror (errno));
		fprintf (stderr, "error: listening to port %d: listen(): %s\n",
				Ident_Port, strerror (errno));
		return;
	}

	fclose (stdin);
	fclose (stderr);
	fclose (stdout);

	syslog (LOG_NOTICE, "identity services started [%s]",
			ID_VERSION);
	d_core (s);

	syslog (LOG_ERR, "error: identity services terminated on "
		"internal error: See last (few) message(s)");

	closelog ();
	close (s);
	return;	
}


define.h

#include "config.h"	/* need this for sys dependancy fun */

/*
 * User the ident service should run as to provide maximum system
 * safety (in case of exploits and such). User nobody should work
 * fine on Linux and FreeBSD. This is only supplied for admins that
 * may perhaps be -ultra- security conscious (insane?) enough to make 
 * a seperate user for ident services. It also allows for fun things
 * with groups.
 */
#define SAFE_ACCESS_USER	"nobody"

/*
 * Port to bind to.
 * I am not aware of any other ident server or ident query agent
 * in existance that uses a port other than 113.
 */
#define IDENT_PORT      113

/*
 * user should create this file in their home directory 
 * if they do not want ident2 to send out ident 
 * replies.
 */
#define	NOIDENT_FILE	".noident"

/*
 * ident server looks for this file in the user's
 * home directory and uses this to decide what to reply
 */
#define USERCONF_FILE   ".ident"

/*
 * number of default max connections (overriden via command-line)
 * that a server will handle at once.
 */
#define	DEF_MAX_CONNECTIONS	25

/*
 * default amount of time (overriden via command-line) to keep
 * any given connection open for (specified in seconds).
 */
#define	DEF_CLIENT_TIMEOUT	20

/*
 * if you happen to use this feature
 */
#define RAND_STRING_LENGTH	6



ident2.c
/* 
 * Ident-2 - an Identity server for UNIX
 * Copyright (C) 1998-2001 Michael Bacarella
 * Copyright (C) 2003 Netgraft Corporation
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 * Please view the file README for program information.
 *
 */

#include "ident2.h"

#ifdef HAS_GETOPT_LONG
#include <getopt.h>	/* getopt() is declared in unistd.h */
#endif

	/*
	 *	determine if we're running as a child of inetd
	 *	or if we were started via a command line (or from
	 *	a process not inetd)
	 */
static int
inetd_child (void)
{
	struct sockaddr_in sin;
	int sinsize = sizeof (struct sockaddr_in);
	
		/* if child of inetd, 0 would be a socket */
	if (getsockname (0, (struct sockaddr *)&sin, (int *)&sinsize) == -1) {
		if (errno == ENOTSOCK)
			return 0;	/* must be a filedescriptor */
		else {
			syslog (LOG_ERR, "getsockname: %s\n", strerror (errno));
			exit (-1);
		}
	}
	return 1;
}

#define print_header()	printf ("Ident2 version %s, Copyright (C) 1999-2004 Michael Bacarella\n", ID_VERSION)

static void
print_usage (void)
{
	print_header();
	puts ("Example command lines:");
puts ("		ident2 -r		sends random replies, always");
puts ("		ident2 -i -y .reply	allows user replies, which are read\n"
      "					from the file ~/.reply");
puts ("		ident2 -m 5 -o 10	ultra secure mode. no more than\n"
      "					5 connections at once, each connection\n"
      "					is killed after 10 seconds regardless.");
puts ("		ident2 -s		stop ident2 from trying to change it's\n"
      "					userid if it discovers that it's running\n"
      "					as a root process.");
}

static void
print_license (void)
{
	print_header();
	puts(
"ident2 is free software, and you are welcome to redistribute it\n"
"under certain conditions; ident2 comes with ABSOLUTELY NO WARRANTY;\n"
"for details, go to http://www.gnu.org/\n");
}

static void
print_help (void)
{
	print_header();
#ifdef HAS_GETOPT_LONG
	puts (  
"usage: ident2 [options]\n"
"options:\n"
"\n"
"these parameters apply to all incarnations of ident2\n"
"\n"
"	-h	--help			this command line information\n"
"	-u	--usage			shows example command lines\n"
"	-v	--version		show version information\n"
"	-l	--license		show licensing information\n"
"	-a	--force-inetd		force inetd mode\n"
"	-d	--force-daemon		force standalone daemon\n"
"	-i	--use-user-ident	allow user defined ident replies\n"
"	-y	--user-reply-file	file in user's homedir for replies\n"
"	-n	--allow-noident-file	don't reply if user has a ~/NOIDENT file\n"
"	-o	--client-timeout	clients timeout after this many secs\n"
"	-s	--dont-change-uid	don't try to change uid (to nobody)\n"
"	-r	--always-random		always send a random reply\n"
"\n"
"these parameters only apply to the daemon ident2\n"
"\n"
"	-m	--daemon-maxclients	accept no more than this many clients\n"
"	-p	--daemon-port		bind this port, instead of 'auth'\n"
);
#else
	puts(
"usage: ident2 [options]\n"
"options:\n"
"\n"
"these parameters apply to all incarnations of ident2\n"
"\n"
"	-h		this command line information\n"
"	-u		shows example command lines\n"
"	-v		show version information\n"
"	-l		show licensing information\n"
"	-a		force inetd mode\n"
"	-d		force standalone daemon\n"
"	-i		allow user defined ident replies\n"
"	-y		file in user's homedir for replies\n"
"	-n		don't reply if user has a ~/NOIDENT file\n"
"	-o		clients timeout after this many secs\n"
"	-s		don't try to change uid (ex: to nobody)\n"
"	-r		always send a random reply, even to bad requests\n"
"\n"
"these parameters only apply to the daemon ident2\n"
"\n"
"	-m		accept no more than this many clients\n"
"	-p		bind this port, instead of 'auth'\n"
);
#endif
}

        /*
         *      print version info
         */
static void
print_version (void)
{
        printf ("Ident2 version %s, maintainer %s\n\n",
                ID_VERSION, ID_MAINTAINER);
}


static void
command_line (int argc, char *argv[])
{
	int c;
			/* getopt_long support implemented
			 * by Alexander Reelsen <ar@rhwd.net> */
#ifdef HAS_GETOPT_LONG
	while (1) {
/* int this_option_optind = optind ? optind : 1; */	/*unused?*/
		int option_index = 0;
		static struct option long_options[] = {
			{"allow-noident-file", 0, 0, 'n'},
			{"help", 0, 0, 'h'},
			{"usage", 0, 0, 'u'},
			{"version", 0, 0, 'v'},
			{"license", 0, 0, 'l'},
			{"force-inetd", 0, 0, 'a'},
			{"force-daemon", 0, 0, 'd'},
			{"use-user-ident", 0, 0, 'i'},
			{"dont-change-uid", 0, 0, 's'},
			{"always-random", 0, 0, 'r'},
			{"user-reply-file", 1, 0, 'y'},
			{"client-timeout", 1, 0, 'o'},
			{"daemon-maxclients", 1, 0, 'm'},
			{"daemon-port", 1, 0, 'p'},
			{0, 0, 0, 0},
		};

		c = getopt_long (argc, argv, "nhuvrliady:o:ftTsm:p:",
			long_options, &option_index);
		if (c == -1) {
			break;
		}	
#else
        while ((c = getopt (argc, argv, "nhuvrliady:o:ftTsm:p:")) != -1) {
#endif
		switch (c) {
			case 'n':
				Allow_NOIDENT = TRUE;
				break;
			case 'h':
				print_help ();
				exit (0);
			case 'u':
				print_usage ();
				exit (0);	
			case 'v':
				print_version ();
				exit (0);
			case 'l':
				print_license ();
				exit (0);
			case 'a':
				Service_Type = INETD;
				break;
			case 'd':
				Service_Type = DAEMON;
				break;
			case 'i':
				Use_User_Ident = TRUE;
				break;
			case 's':
				Dont_Change_Uid = TRUE;
				break;
			case 'r':
				Reply_Always_Random = TRUE;
				break;
			case 'y':
				User_Ident_File = optarg;
				break;
			case 'o':
				if ((Client_Timeout = atol (optarg)) == 0) {
					fprintf (stderr, "ERROR: Ident2: bad "
						"value for arguement `o'\n");
					exit (1);
				}
				break;
			case 'm':
				if ((Max_Connections = atol (optarg)) == 0) {
					fprintf (stderr, "ERROR: Ident2: bad"
						" value for arguement `m'\n");
					exit (1);
				}
				break;
			case 'p':
				if ((Ident_Port = atoi (optarg)) == 0) { 
					fprintf (stderr, "ERROR: Ident2: bad"
						" value for arguement `p'\n");
					exit (1);
				}
				break;
			case '?':
			default:
#ifdef HAS_GETOPT_LONG
				fprintf (stderr, "ident2: bad command line:\n"
				"\t``ident2 --help | more'' for help\n"
				"\t``ident2 --usage'' for some examples\n\n");
#else
				fprintf (stderr, "ident2: bad command line:\n"
				"\t``ident2 -h | more'' for help\n"
				"\t``ident2 -u'' for some examples\n\n");
#endif
				exit (1);
		}
	}
}


int
main (int argc, char *argv[])
{	
	INIT_GLOBALS();
	command_line (argc, argv);
	 	
			/* try to determine the service type if nothing
			 * has been forced (by commandline) */
	if (Service_Type == NO_TYPE) {
		if (inetd_child ())
			Service_Type = INETD;
		else
			Service_Type = DAEMON;
	}

	if (Service_Type == INETD) {
		openlog ("in.ident2", LOG_PID, LOG_AUTH);
		child_service (STDIN_FILENO, STDOUT_FILENO);
		closelog ();
	}
	else
		daemon_service ();

	return 0;
}

